hypertrace-view-generator/hypertrace-view-generator/src/main/java/org/hypertrace/viewgenerator/generators/ViewGeneratorState.java

```diff
@@ -0,0 +1,95 @@
+package org.hypertrace.viewgenerator.generators;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.hypertrace.core.datamodel.Entity;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.EventRef;
+import org.hypertrace.core.datamodel.EventRefType;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.traceenricher.trace.util.ApiTraceGraph;
+
+public class ViewGeneratorState {
+
+  private static final ThreadLocal<TraceState> traceStateThreadLocal = new ThreadLocal<>();
+  private static final ThreadLocal<ApiTraceGraph> apiTraceGraphThreadLocal = new ThreadLocal<>();
+
+  public static ApiTraceGraph getApiTraceGraph(StructuredTrace trace) {
+    if (apiTraceGraphThreadLocal.get() == null
+        || isDifferentTrace(apiTraceGraphThreadLocal.get().getTrace(), trace)) {
+      apiTraceGraphThreadLocal.set(new ApiTraceGraph(trace).build());
+    }
+    return apiTraceGraphThreadLocal.get();
+  }
+
+  public static TraceState getTraceState(StructuredTrace trace) {
+    if (traceStateThreadLocal.get() == null
+        || isDifferentTrace(traceStateThreadLocal.get().getTrace(), trace)) {
+      traceStateThreadLocal.set(new TraceState(trace));
+    }
+    return traceStateThreadLocal.get();
+  }
+
+  private static boolean isDifferentTrace(StructuredTrace cached, StructuredTrace trace) {
```

Comment:
When will this happen in view generators?

Reply:
When the same trace is getting processed by multiple view-generators, this would return false, and we save on some repeated operations.
When a new trace starts getting processed, this would return false, leading to rebuilding of TraceState and ApiTraceGraph

Reply:
I was mainly concerned/curious about the customer id check, but I guess that's just done for safety? Because, a trace being different means, a different trace id

Reply:
yes, just an additional check

Reply:
I think checking for traceId is also not required, since the input trace doesn't get modified in this service. So all we need is to check for same object.